services/navigation/builtin/builtin.go

```diff
@@ -220,90 +218,52 @@ func (svc *builtIn) SetMode(ctx context.Context, mode navigation.Mode, extra map
+			svc.logger.Errorw("failed to get gps location", "error", err)
+		}
+
+		navOnce := func(ctx context.Context) error {
+			bearingToGoal, _, err := svc.waypointDirectionAndDistanceToGo(ctx, currentLoc)
 			if err != nil {
-				svc.logger.Errorw("failed to get gps location", "error", err)
-				continue
+				return err
 			}
 
-			if len(path) <= 1 || currentLoc.GreatCircleDistance(path[len(path)-1]) > .0001 {
-				// gps often updates less frequently
-				path = append(path, currentLoc)
-				if len(path) > 2 {
-					path = path[len(path)-2:]
-				}
+			wp, err := svc.nextWaypoint(ctx)
+			if err != nil {
+				return fmt.Errorf("waypoint error: %w", err)
 			}
 
-			navOnce := func(ctx context.Context) error {
-				if len(path) <= 1 {
-					return errors.New("not enough gps data")
-				}
-
-				currentBearing, err := svc.computeCurrentBearing(ctx, path)
-				if err != nil {
-					return err
-				}
-
-				bearingToGoal, distanceToGoal, err := svc.waypointDirectionAndDistanceToGo(ctx, currentLoc)
-				if err != nil {
-					return err
-				}
-
-				if distanceToGoal < .005 {
-					svc.logger.Debug("i made it")
-					return svc.waypointReached(ctx)
-				}
-
-				bearingDelta := computeBearing(bearingToGoal, currentBearing)
-				steeringDir := -bearingDelta / 180.0
+			goal := wp.ToPoint()
 
-				svc.logger.Debugf("currentBearing: %0.0f bearingToGoal: %0.0f distanceToGoal: %0.3f bearingDelta: %0.1f steeringDir: %0.2f",
-					currentBearing, bearingToGoal, distanceToGoal, bearingDelta, steeringDir)
+			// have ability to define destination heading here, but waypoint structure doesn't allow for that so using bearingToGoal as heading
+			_, err = svc.motion.MoveOnGlobe(ctx, svc.base.Name(), goal, bearingToGoal, svc.movementSensor.Name(), svc.obstacles, svc.metersPerSec*1000, svc.degPerSec, extra)
 
-				// TODO(erh->erd): maybe need an arc/stroke abstraction?
-				// - Remember that we added -1*bearingDelta instead of steeringDir
-				// - Test both naval/land to prove it works
-				if err := svc.base.Spin(ctx, -1*bearingDelta, svc.degPerSec, nil); err != nil {
-					return fmt.Errorf("error turning: %w", err)
-				}
-
-				distanceMm := distanceToGoal * 1000 * 1000
-				distanceMm = math.Min(distanceMm, 10*1000)
+			if err == nil {
+				return svc.waypointReached(ctx)
+			}
 
-				// TODO: handle swap from mm to meters
-				if err := svc.base.MoveStraight(ctx, int(distanceMm), (svc.metersPerSec * 1000), nil); err != nil {
-					return fmt.Errorf("error moving %w", err)
-				}
+			return err
+		}
 
-				return nil
-			}
+		_, err = svc.nextWaypoint(svc.cancelCtx)
+		if err != nil {
+			svc.logger.Infof("waypoint error: %w", err)
+		}
 
-			if err := navOnce(svc.cancelCtx); err != nil {
-				svc.logger.Infof("error navigating: %s", err)
+		// loop until no waypoints remaining
+		for err == nil {
+			if navErr := navOnce(svc.cancelCtx); err != nil {
+				svc.logger.Infof("error navigating: %s", navErr)
 			}
+			_, err = svc.nextWaypoint(svc.cancelCtx)
```

Comment:
[q] what is the purpose of doing this if you are not looking at the return value?

Reply:
it was for the next iteration of the for loop so that the error could be checked to determine if there are any more waypoints to visit. I changed this in my new commit to be cleaner.

Reply:
Looks a lot cleaner nice!